### PR TITLE
[REVIEW] Fix inner_join incorrect result issue

### LINKED ISSUE #3702: [BUG] innerJoin in libcudf++ can produce incorrect results
**Describe the bug**
innerJoin can sometimes replicate the first column at the end of a result instead of including all of the results. The number of results is always correct, but the contents appears to be wrong some times, and I don't know what is triggering it.

**Steps/Code to reproduce bug**
```diff
diff --git a/cpp/tests/join/join_tests.cu b/cpp/tests/join/join_tests.cu
index 675910e44..1dddad644 100644
--- a/cpp/tests/join/join_tests.cu
+++ b/cpp/tests/join/join_tests.cu
@@ -222,6 +222,33 @@ TEST_F(JoinTest, LeftJoinWithNulls)
   cudf::test::expect_tables_equal(*sorted_gold, *sorted_result);
 }
 
+
+TEST_F(JoinTest, InnerJoinProblem)
+{
+  column_wrapper <int64_t> col0_0{{4, 1, 3, 2, 2, 2, 2}};
+  column_wrapper <int64_t> col1_0{{2}};
+
+  CVector cols0, cols1;
+  cols0.push_back(col0_0.release());
+  cols1.push_back(col1_0.release());
+
+  Table t0(std::move(cols0));
+  Table t1(std::move(cols1));
+
+  auto result = cudf::experimental::inner_join(t0, t1, {0}, {0}, {{0, 0}});
+  auto result_sort_order = cudf::experimental::sorted_order(result->view());
+  auto sorted_result = cudf::experimental::gather(result->view(), *result_sort_order);
+
+  column_wrapper <int64_t> col_gold_0{{2, 2, 2, 2}};
+  CVector cols_gold;
+  cols_gold.push_back(col_gold_0.release());
+  Table gold(std::move(cols_gold));
+
+  auto gold_sort_order = cudf::experimental::sorted_order(gold.view());
+  auto sorted_gold = cudf::experimental::gather(gold.view(), *gold_sort_order);
+  cudf::test::expect_tables_equal(*sorted_gold, *sorted_result);
+}
+
 TEST_F(JoinTest, InnerJoinNoNulls)
 {
   column_wrapper <int32_t> col0_0{{3, 1, 2, 0, 2}};
```

I have a longer version with more columns/rows, but I got it down to this after some trial and error.

**Expected behavior**
Test should pass, but it fails.

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
expected_size was getting changed even before it was getting compared because of which the indices size found were smaller and resizing later updated rest of the values with 0's.

closes #3702

### CODE REVIEW

File: cpp/tests/join/join_tests.cu [hunk 1]
```diff
@@ -584,3 +584,29 @@ TEST_F(JoinTest, EqualValuesFullJoin)
 
   cudf::test::expect_tables_equal(gold, *result);
 }
+
+TEST_F(JoinTest, InnerJoinCorenerCase)
```
Comment (kaatish) [inline cpp/tests/join/join_tests.cu]: InnerJoinCorenerCase > InnerJoinCornerCase

**Comment by revans2** [comment]
This fixes the issues that I was seeing.

File: cpp/src/join/hash_join.cuh [hunk 2]
```diff
@@ -322,8 +322,7 @@ get_base_hash_join_indices(
     CHECK_CUDA(stream);
 
     join_size = write_index.value();
-    estimated_size *= 2;
-  } while (estimated_size < join_size) ;
+  } while ((estimated_size < join_size) and (estimated_size *= 2)) ;
```
Comment (jrhemstad) [inline cpp/src/join/hash_join.cuh]: This is potentially confusing. Mixing an assignment inside a logic check isn't very clear. Can this be done another way?

Comment (rgsl888prabhu) [reply cpp/src/join/hash_join.cuh]: I can create one more variable like current_estimated_size like below
```
current_estimated_size = estimated_size;
do {
...........
...........
current_estimated_size = estimated_size;
estimated_size *= 2;
} while (current_estimated_size < join_size) ;


```

Comment (rgsl888prabhu) [reply cpp/src/join/hash_join.cuh]: Or better way 
```
do {
if (join_size != 0){
estimated_size *= 2;
}
...........
...........
}while (estimated_size < join_size) ;
```

Comment (rgsl888prabhu) [reply cpp/src/join/hash_join.cuh]: moving with option 1 as it will not be dependent on join_size.

**Comment by rgsl888prabhu** [comment]
rerun test

**Comment by rgsl888prabhu** [comment]
rerun tests

**Pull request merged**